webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala

```diff
@@ -90,4 +96,98 @@ class ReadOnlySoftwareDAOImpl(
       softwareByNode.mapValues { ids => software.filter(s => ids.contains(s.id)) }
     }
   }
+
+  def getAllSoftwareIds() : Box[Set[SoftwareUuid]] = {
+    for {
+      con     <- ldap
+      softwareEntry = con.searchOne(inventoryDitService.getSoftwareBaseDN, IS(OC_SOFTWARE), A_SOFTWARE_UUID)
+      ids     <- sequence(softwareEntry) { entry =>
+        entry(A_SOFTWARE_UUID) match {
+          case Some(value) => Full(value)
+          case _ => Failure(s"Missing attribute ${A_SOFTWARE_UUID} for entry ${entry.dn} ${entry.toString()}")
+        }
+      }
+      softIds       =  ids.map(id => SoftwareUuid(id)).toSet
+    } yield {
+      softIds
+    }
+  }
+
+  def getSoftwaresForAllNodes() : Box[Set[SoftwareUuid]] = {
+    // fetch all softwares, for all nodes, in all 3 dits
+    val acceptedDit = inventoryDitService.getDit(AcceptedInventory)
+
+    var mutSetSoftwares: Box[scala.collection.mutable.Set[SoftwareUuid]] = Full(scala.collection.mutable.Set[SoftwareUuid]())
+
+    val t1 = System.currentTimeMillis
+    for {
+      con           <- ldap
+
+      // fetch all nodes
+      nodes         = con.searchSub(acceptedDit.NODES.dn.getParent, IS(OC_NODE), A_NODE_UUID)
+
+      batchedNodes = nodes.grouped(50)
+
+      _ = batchedNodes.foreach { nodeEntries: Seq[LDAPEntry] =>
```

Comment:
it should not be a foreach, else we loose the maybe `Failure`:  ```res <- sequence(batchedNodes) { .... }```

Reply:
i tried using a sequence on batchedNodes, without success (it's an iterator, I cannot sequence it)

Reply:
You can do:
``
      batchedNodes         = con.searchSub(acceptedDit.NODES.dn.getParent, IS(OC_NODE), A_NODE_UUID).grouped(50).toSeq

      res <- sequence(batchedNodes) { nodeEntries: Seq[LDAPEntry] =>
```
You still process nodes 50 by 50, and don't hold more node ids than previously